event-sourcing/subscription-manager/src/main/java/uk/gov/justice/services/event/sourcing/subscription/startup/manager/ConcurrentEventStreamConsumerManager.java

```diff
@@ -0,0 +1,73 @@
+package uk.gov.justice.services.event.sourcing.subscription.startup.manager;
+
+import static java.lang.String.format;
+
+import uk.gov.justice.services.event.sourcing.subscription.startup.EventCatchupException;
+import uk.gov.justice.services.event.sourcing.subscription.startup.listener.EventStreamConsumerListener;
+import uk.gov.justice.services.event.sourcing.subscription.startup.listener.FinishedProcessingMessage;
+import uk.gov.justice.services.event.sourcing.subscription.startup.task.ConsumeEventQueueTask;
+import uk.gov.justice.services.event.sourcing.subscription.startup.task.ConsumeEventQueueTaskFactory;
+import uk.gov.justice.services.messaging.JsonEnvelope;
+
+import java.util.Queue;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import javax.enterprise.concurrent.ManagedExecutorService;
+
+public class ConcurrentEventStreamConsumerManager implements EventStreamConsumerManager, EventStreamConsumerListener {
+
+    private static final Object MUT_EX = new Object();
+
+    private final ConcurrentHashMap<UUID, Queue<JsonEnvelope>> allEventStreams = new ConcurrentHashMap<>();
+    private final ConcurrentLinkedQueue<Queue<JsonEnvelope>> eventStreamsInProgress = new ConcurrentLinkedQueue<>();
+
+    private final ManagedExecutorService managedExecutorService;
+    private final ConsumeEventQueueTaskFactory consumeEventQueueTaskFactory;
+
+    public ConcurrentEventStreamConsumerManager(final ManagedExecutorService managedExecutorService,
+                                                final ConsumeEventQueueTaskFactory consumeEventQueueTaskFactory) {
+        this.managedExecutorService = managedExecutorService;
+        this.consumeEventQueueTaskFactory = consumeEventQueueTaskFactory;
+    }
+
+    @Override
+    public int add(final JsonEnvelope event) {
+
+        final UUID streamId = event.metadata()
+                .streamId()
+                .orElseThrow(() -> missingStreamIdException(event.metadata().id()));
+
+        final Queue<JsonEnvelope> events = allEventStreams.computeIfAbsent(streamId, id -> new ConcurrentLinkedQueue<>());
+
+        synchronized (MUT_EX) {
+            events.offer(event);
+
+            if (shouldCreateTask()) {
+                eventStreamsInProgress.add(events);
+
+                final ConsumeEventQueueTask consumeEventQueueTask = consumeEventQueueTaskFactory.createWith(events, this);
+
+                managedExecutorService.submit(consumeEventQueueTask);
+            }
+        }
+
+        return 1;
+    }
+
```

Comment:
Need a method called after all events are added, that will check that all the queues are empty or currently being processed.